File: solutions/0039-combination-sum/solution.cpp
class Solution {
public:
    void backtrack(std::vector<int>& candidates, int need, int index, std::vector<int> cur, std::vector<std::vector<int>>& result) {
        if (need <= 0) {
            if (!need) {
                result.push_back(cur);
            }
            return;
        }
        for (int i = index; i < candidates.size(); ++i) {
            cur.push_back(candidates[i]);
            backtrack(candidates, (need - candidates[i]), i, cur, result);
            cur.pop_back();
        }
    }

    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> l_ret;
        if (candidates.empty()) return l_ret;
        if (candidates.size() == 1) {
            if (target == candidates[0]) {
                l_ret.push_back({candidates[0]});
            }
            return l_ret;
        }
        backtrack(candidates, target, 0, {}, l_ret);
        return l_ret;
    }
};

File: solutions/0039-combination-sum/solution.go
// Output below for learning
func combinationSum(nums []int, target int) [][]int {
    if len(nums) == 0 { return [][]int{} }
	if len(nums) == 1 {
		if target == nums[0] {
			return [][]int{{nums[0]}}
		}
		return [][]int{}
	}
	var backtrack func(need int, index int)
	var allResults [][]int
	var currentPath []int
	backtrack = func(need int, index int) {
		// fmt.Println(currentPath)
		if need == 0 {
			res := make([]int, len(currentPath))
			copy(res, currentPath)
			// fmt.Print("Selected - ")
			// fmt.Println(res)
			allResults = append(allResults, res)
			// currentPath = currentPath[:0]
		} else if need < 0 {
			// Has moved past, so need to retrack
			return
		}
		for i := index; i < len(nums); i++ {
			currentPath = append(currentPath, nums[i])
			backtrack(need - nums[i], i)
			currentPath = currentPath[0:(len(currentPath)-1)]
		}
	}
	backtrack(target, 0)
	return allResults
}

/*
[]
[2]
[2 2]
[2 2 2]
[2 2 2 2]
[2 2 2 2 2]
[2 2 2 2 5]
[2 2 2 2 6]
[2 2 2 2 9]
[2 2 2 5]
[2 2 2 6]
[2 2 2 9]
[2 2 5]
Selected - [2 2 5]
[2 2 5 5]
[2 2 5 6]
[2 2 5 9]
[2 2 6]
[2 2 9]
[2 5]
[2 5 5]
[2 5 6]
[2 5 9]
[2 6]
[2 6 6]
[2 6 9]
[2 9]
[5]
[5 5]
[5 6]
[5 9]
[6]
[6 6]
[6 9]
[9]
Selected - [9]
[9 9]

*/
